hilti/runtime/include/util.h

```diff
@@ -649,4 +650,97 @@ constexpr std::size_t hashCombine(std::size_t hash1, std::size_t hash2, Hashes..
         return result;
 }
 
+template<typename Data, typename Error>
+class ControlBlockRef;
+
+// Helper class for loosely tracking liveliness of some memory. Code using this
+// would instantiate an instance of this pointing to some data. It can then
+// pass out references to that control block by calling `asRef`. One can get
+// values from these references as long as the original control block is live.
+template<typename Data, typename Error>
+class ControlBlock {
+public:
+    using Ref = ControlBlockRef<Data, Error>;
+
+    ControlBlock() = default;
+    ControlBlock(Data* data) : _data(data) {}
+
+    ControlBlock(const ControlBlock&) = default;
+    ControlBlock(ControlBlock&&) = default;
+
+    ControlBlock& operator=(const ControlBlock&) = default;
+    ControlBlock& operator=(ControlBlock&&) = default;
+
+    friend bool operator==(const ControlBlock& a, const ControlBlock& b) {
+        return std::tie(a._control, a._data) == std::tie(b._control, b._data);
+    }
+
+    friend bool operator!=(const ControlBlock& a, const ControlBlock& b) { return ! (a == b); }
+
+    /*implicit*/ operator Ref() const {
+        if ( ! _control )
+            _control = std::make_shared<bool>();
+
+        return {_control, _data};
+    }
+
+    // Invalidate all references to this control block.
+    void reset() { _control.reset(); }
+
+private:
+    mutable std::shared_ptr<void> _control;
+    Data* _data = nullptr;
+};
+
+
+template<typename Data, typename Error>
+class ControlBlockRef {
+public:
+    ControlBlockRef() = default;
+    ControlBlockRef(const ControlBlockRef&) = default;
+    ControlBlockRef(ControlBlockRef&&) = default;
+
+    ControlBlockRef& operator=(const ControlBlockRef&) = default;
+    ControlBlockRef& operator=(ControlBlockRef&&) = default;
+
+    bool isValid() const { return _data && ! _control.expired(); }
+
+    // Get reference to the controlled object. Throws if the object has expired or the data is invalid..
+    const Data& get() const {
+        if ( ! _data )
+            throw Error("bound object is invalid");
```

Comment:
```suggestion
            throw Error("underlying object is invalid");
```

As this is potentially user-facing (more instances below).